### PR TITLE
Fix stack overflow on assembly resolution

### PR DESCRIPTION
**Pull request by KevinRansom**
Well .... this is embarrassing ... as I have said before our build is dependent on a very old build of the coreclr for testing.

We had a bug based on a badly implemented LoadContext (Sorry about that ... all my fault), that works on this very old coreclr but was fixed soon after ... only .... well  we couldn't update at that time.  On a fixed coreclr this old bad but working code becomes a stack overflow.  Because I believed it only impacted fsi.exe, and there is a lot more I want to do for that tool I wasn't in any rush to fix it --- bad descision because it also impacts TypeProviders which I forgot about.

Anyway ... because we still need to use the old coreclr for a bit longer ... only for testing of course ...  the fix here tests to see if the mscorlib is of a very specific version ... **the one we use for testing**.  We do the old thing ... if instead it is any other version in existence we do the right thing.

Anyway please don't hate on me ... I am doing plenty of that myself.  I wish that I was smart enough to come up with a good excuse that blamed others but this one was all on me ...

Fixes: #3658

### CODE REVIEW

**Comment by KevinRansom** [comment]
@dsyme @xenocons , do either of you have a repro to test.  This fixes the issue that I know of but that one doesn't use a TP.

Thanks

Kevin

**Comment by xenocons** [comment]
Hi Kevin, We all have bugs :) 

I have difficulties to compile fsharp on netcore 2 with debian at the moment, however as a repo:

```
dotnet /usr/share/dotnet/sdk/2.0.0/FSharp/fsc.exe --noframework -r:/usr/share/dotnet/sdk/NuGetFallbackFolder/netstandard.library/2.0.0/build/netstandard2.0/ref/mscorlib.dll -r:/usr/share/dotnet/sdk/NuGetFallbackFolder/netstandard.library/2.0.0/build/netstandard2.0/ref/System.Runtime.dll -r:/usr/share/dotnet/sdk/NuGetFallbackFolder/netstandard.library/2.0.0/build/netstandard2.0/ref/netstandard.dll -r:/home/xenocons/.nuget/packages/sqlprovider/1.1.11/lib/FSharp.Data.SqlProvider.dll test.fs
```

`test.fs` can be anything, and the bug occurs with the inclusion of sqlprovider (you will need to modify the path in the above command). For anyone else reading this links to https://github.com/Microsoft/visualfsharp/issues/3654

File: src/utils/reshapedreflection.fs [hunk 1]
```diff
@@ -48,6 +48,10 @@ module internal ReflectionAdapters =
     let isStaticFlag    f    = hasFlag BindingFlags.Static f
     let isInstanceFlag  f    = hasFlag BindingFlags.Instance f
     let isNonPublicFlag f    = hasFlag BindingFlags.NonPublic f
+ 
+    let isOldCorlib =
+        let attributes = typeof<RuntimeTypeHandle>.GetTypeInfo().Assembly.GetCustomAttributes<AssemblyFileVersionAttribute>() |> Seq.toArray
+        if attributes.Length = 0 then false else attributes.[0].Version = "4.6.24410.01"
```
Comment (forki) [inline src/utils/reshapedreflection.fs]: let attributes = typeof<RuntimeTypeHandle>.GetTypeInfo().Assembly.GetCustomAttributes<AssemblyFileVersionAttribute>()
    match Seq.tryHead attributes with
    | Some a -> a.Version = "4.6.24410.01"
    | _ -> false

Comment (forki) [reply src/utils/reshapedreflection.fs]: also I assume it should be moved down next to usage

Comment (forki) [reply src/utils/reshapedreflection.fs]: and the the bool you can remove the temporary bool completely ....

**Comment by dsyme** [comment]
Great to have this fixed. If it unblocks development of type providers for .NET Core I'll be very very happy

**Comment by KevinRansom** [comment]
Okay this addresses the stack overflow report.  However it moves the cheese down the line a bit:

There are two further issues from this:

1. when building f# with the dotnet/cli f# somehow ... I don't know how picks the dotnet/cli from the path.
````
c:\temp\repro>c:\microsoft\visualfsharp\Tools\dotnet20\dotnet.exe build /v:normal

   ... a ton of logging elided

         C:\Program Files\dotnet\sdk\2.0.2-vspre-006949\FSharp\RunFsc.cmd -o:obj\Debug\netcoreapp2.0\repro.dll
      
````
2. ProvidedType definition doesn't implement IReflect.

````
C:\microsoft\visualfsharp\tests\testbin\release\coreclr\FSC>C:\microsoft\visualfsharp\Tools\dotnet20\dotnet C:\microsoft\visualfsharp\Tools\dotnet20\sdk\2.0.0-preview2-006502\FSharp\fsc.exe @c:\temp\repro\repro.rsp
Microsoft (R) F# Compiler version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.

error FS3033: The type provider 'FSharp.Data.Sql.SqlTypeProvider' reported an error: Unable to cast object of type 'ProviderImplementation.ProvidedTypes.ProvidedTypeDefinition' to type 'System.Reflection.IReflectableType'.

error FS3020: One or more errors seen during provided type setup

````

**Comment by KevinRansom** [comment]
1. is a dotnet sdk issue not ours :-)

**Comment by KevinRansom** [comment]
@dotnet-bot test Windows_NT Release_ci_part2 Build  please

**Comment by xenocons** [comment]
@KevinRansom could you confirm if the point 2. that you raised above would be remediated by implementing IReflectableType on https://github.com/fsprojects/FSharp.TypeProviders.StarterPack ?

**Comment by KevinRansom** [comment]
@xenocons 

I have been looking at it and am kind of confused.  It is possible to GetTypeinfo() from a Type despite Type not implementing IReflectableType.  So I need to understand why the runtime is failing to do it.

On CoreClr type doesn't implement IReclectableType.

CoreClr type

image

Anyway ... working on it ...

**Comment by KevinRansom** [comment]
@dsyme we can't implement IReflectableType because TypeInfo does not have a public constructor.  I think this may be a reference assemblies thing.  But I am still worrying about p2p refs right now.

**Comment by KevinRansom** [comment]
@dsyme, @brettfo, @cartermp, @Pilchie 

I had another chat with Immo.  This is clearly a bug in the framework and it is unclear how it can be fixed.  Type extensibility by deriving from Type is incompatible with reflection using TypeInfo.

- The most correct fix would be to make the TypeInfo() constructor protected or public, allowing us to implement a TypeInfo() which would then delegate to the generated type.  However, there exists no dotnet runtime anywhere with that public, so it would be somewhat limited in terms of reach.

- What we will do is this ... build the compiler for netcoreapp2.0 and ensuring that we don't use GetTypeInfo().  This will mean that either ...
we enable build for both netcoreapp2.0 and FSharp.Core 1.0 or we stop shipping with dotnet/cli 1.1.*

Given that VS now has a dotnet core 2.0 install, I doubt the dotnet core 1.0 release matters much to us now.

Obviously all of the usual Kevin is embarrassed about this excuses spring to the fore here.  I don't believe we use GetTypeInfo in FSharp.Core so I think we can stick with a NetStandard 1.6 FSharp.Core.dll

**Pull request merged**